src/dbnode/persist/fs/commitlog/commit_log.go

```diff
@@ -506,69 +541,169 @@ func (l *commitLog) write() {
 		l.metrics.success.Inc(numWritesSuccess)
 	}
 
-	writer := l.writerState.writer
-	l.writerState.writer = nil
+	// Ensure that there is no active background goroutine in the middle of reseting
+	// the secondary writer / modifying its state.
+	l.waitForSecondaryWriterAsyncResetComplete()
+	// Can be nil in the case where the background goroutine spawned in openWriters
+	// encountered an error trying to re-open it.
+	if l.writerState.secondaryWriter.writer != nil {
+		// Don't care about errors closing the secondary writer because it doesn't
+		// have any data.
+		l.writerState.secondaryWriter.writer.Close()
+		l.writerState.secondaryWriter.writer = nil
+	}
+
+	writer := l.writerState.primaryWriter.writer
+	l.writerState.primaryWriter.writer = nil
 
 	l.closeErr <- writer.Close()
 }
 
-func (l *commitLog) onFlush(err error) {
-	l.flushState.setLastFlushAt(l.nowFn())
+// newOnFlushFn is used to create new flushFns because each one needs to know which
+// slice of pendingFlushFns it should modify.
+func (l *commitLog) newOnFlushFn(writer *asyncResettableWriter) flushFn {
```

Comment:
Hm, why don't you add the `onFlush(err error)` method to `asyncResettableWriter` and then make it call the commit log back.

i.e.

```go
type asyncResettableWriter struct {
  // .. other
  commitLog *commitLog
}

func (w *asyncResettableWriter) onFlush(err error) {
  w.commitLog.onFlush(w, err)
}

// now on commit log, retain how this used to work
func (l *commitLog) onFlush(writer *asyncResettableWriter, err error) {
  // as before
}
```

Reply:
Yeah I thought about this and didn't do it because i really hate cyclic datastructures as they can be really confusing. Probably simpler to do it in this case though, I'll make the change